src/main/client/network/buffered-network-component.ts

```diff
@@ -0,0 +1,87 @@
+import { Singleton } from 'typescript-ioc';
+import { NetworkEvent, NetworkMessage } from './network-model';
+import { SharedConfig } from '../../shared/config/shared-config';
+import deepmerge from 'deepmerge';
+import { IObject } from '../../shared/util/util-model';
+import { spawn, Worker } from 'threads/dist';
+import { NetworkThread } from './network-thread/network-thread';
+import 'threads/register';
+import { Subject } from 'rxjs';
+
+@Singleton
+export class BufferedNetworkComponent {
+   private readonly bufferTimerMs = 1000 / SharedConfig.NETWORK_TICK_RATE;
+   private bufferedEventsMessages = new Map<NetworkEvent, IObject>();
+   private lastSendTime = 0;
+   private readonly bindRequestBufferTimer;
+   private sending = false;
+   private networkThread?: NetworkThread;
+
+   private readonly connectedSubject = new Subject<void>();
+   readonly connected$ = this.connectedSubject.pipe();
+   private readonly disconnectedSubject = new Subject<void>();
+   readonly disconnected$ = this.disconnectedSubject.pipe();
+   private readonly dataSubject = new Subject<NetworkMessage[]>();
+   readonly data$ = this.dataSubject.pipe();
+
+   constructor() {
+      this.bindRequestBufferTimer = this.requestBufferTimer.bind(this);
+      this.initNetworkThread();
+   }
+
+   private async initNetworkThread(): Promise<void> {
+      this.networkThread = ((await spawn(
+         new Worker('./network-thread/network-thread', { type: 'module' }),
+      )) as unknown) as NetworkThread;
+      this.networkThread.onConnected().subscribe(() => this.connectedSubject.next());
+      this.networkThread.onDisconnected().subscribe(() => this.disconnectedSubject.next());
+      this.networkThread.onData().subscribe((data) => this.dataSubject.next(data));
+      this.requestBufferTimer();
+   }
+
+   send(event: NetworkEvent, value: IObject): void {
+      const mergedMessage = deepmerge.all([this.getEventMessage(event), value]) as IObject;
+      this.bufferedEventsMessages.set(event, mergedMessage);
+      this.sendBufferedEventMessagesInTime();
+   }
+
+   private requestBufferTimer(): void {
+      this.sendBufferedEventMessagesInTime();
+      requestAnimationFrame(this.bindRequestBufferTimer);
+   }
+
+   protected sendBufferedEventMessagesInTime(): void {
+      if (!this.sending && Date.now() > this.lastSendTime + this.bufferTimerMs) {
+         this.sendBufferedEvents();
+      }
+   }
+
+   private async sendBufferedEvents(): Promise<void> {
```

Comment:
is it good to buffer network events? What are the advantages?

Reply:
The name might not self explanatory but it does not gather positions and sends it in a group but holds only the newest value and sends that one. By doing that the clients will not send out tremendous data to the server and to other clients unnecessarily. By my experience, a tick rate of 8 is enough. It means the client will send its newest position maximum 8 times per sec. It is good enough to save computing power but fast enoguh to avoid lags on other clients by using interpolation.
Do you have better name for this? I know it is not the best.

Reply:
The way I started to split network components might also not the best way. I have NetworkComponent, StoreNetworkComponent, BufferedNetworkComponent, NetworkThread and SocketIoWrapper. We should discuss it at some point what you think is the best way manage the responsibilities/relationship between those. It can wait though, not the most important is to make a super clean component structure but we need to make progress in the gameplay. Or at least I need, you alreade made a lot! :D